action.yml

```diff
@@ -220,7 +221,7 @@ runs:
           ${{ inputs.stack-arguments }}
 
     - name: Save dependencies cache
-      if: steps.restore-deps.outputs.cache-hit != 'true'
+      if: success() || (failure() && inputs.cache-save-always)
```

Comment:
Sorry, don't merge away yet.

I think inputs are always strings. So `"false"` will still come out as truthy. I think you want to use `== 'true'` or `!= 'false'` here to get the behavior you want.

Should we add a test case?

Reply:
I think you're right. A test case added to `example.yml` is a good call, but testing the input seems like it requires either introducing another example stack project so that a failure occurs, or somehow triggering a failure in the existing project.

Since I would like to adopt this input in `megarepo`, I'll start a branch to use this input and change something in the project to demonstrate the build failing.

Reply:
Testing this input here: https://github.com/freckle/megarepo/pull/32424

Reply:
The [CI logs][0] show the `actions/cache/save` occurring after `stack build` fails.

[0]: https://github.com/freckle/megarepo/actions/runs/7589347043/job/20673802594?pr=32424

Reply:
> The [CI logs](https://github.com/freckle/megarepo/actions/runs/7589347043/job/20673802594?pr=32424) show the actions/cache/save occurring after stack build fails.

Cool, but I think the important thing to test is that it *doesn't* happen when you don't configure save-always. That would be the bug that treating `"false"` a a truthy non-empty string would cause. Is there an example of that you can link to?

Reply:
I think this expression must be erroneous because if these were actually booleans and not strings, the existing `!= 'true'` conditional would always fail...

Reply:
> if these were actually booleans and not strings, the existing != 'true' conditional would always fail...

Why do you expect that? I think they are actually strings, not booleans.

Reply:
I'm saying your suspicion about this being incorrect is true, they should indeed be strings.

Reply:
Oh OK.

Reply:
I think my inability to see a full cache-hit occur is unrelated to this PR or the repro producing those earlier CI runs. In this repo, I've looked at `test (ubuntu-latest, lts)`. After a successful initial run on d36a2d5, I looked at the cache key chosen for the main `stack build` (not dependencies):

- [Retry 1][0] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-817889aa22ef22a7891ffa6f041b5001f4b769f7e747e28c8b339a918e098250`
- [Retry 2][1] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-a8d203ecdad23a840690f55fc64fe52997690acbfc22aef34cc023058d70a634`
- [Retry 3][2] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-7cb01ff854833118b39dcd768d6e7bde63da2c3e9b68539977ab778763a5c347`
- [Retry 4][3] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-0bcad1e2578b1528b1af397a3419c282db865812a77db209e9b189bd5cd2f994`

Each `key` is the same except for the last hash, which is `steps.setup.outputs.sources-hash`. This is computed as:

```bash
echo "sources-hash=${{ hashFiles('**', '!**/.stack-work') }}" >>"$GITHUB_OUTPUT"
```

I have no idea why this would change from run to run, but I am pretty sure this PR is not introducing this behavior and that it should be sorted out separately.

[0]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20785923748?pr=34
[1]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786256605?pr=34
[2]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786323717?pr=34
[3]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786376797?pr=34

Reply:
> I have no idea why this would change from run to run

Is the `hashFiles` call picking up content in `.git`, which will change with every commit?

Could add another `'!./.git'` pattern there, but I agree it's something we can figure out separately.